feat(signals): allow user-defined signals in `withState` and `signalState` by splitting `STATE_SOURCE`

rainerhahnekamp opened the pull request:
BREAKING CHANGES:

`withState` and `signalState` now support user-defined signals like
`linkedSignal`, `resource.value`, or any other `WritableSignal`.

For example:

```ts
const user = signal({ id: 1, name: 'John Doe' });
const userClone = linkedSignal(user);
const userValue = resource({
  loader: () => Promise.resolve('user'),
  defaultValue: ''
});

const Store = signalStore(
  withState({ user, userClone, userValue: userValue.value })
);
```

The state slices don't change:

```ts
store.user;       // DeepSignal<{ id: number, name: string }>
store.userClone;  // DeepSignal<{ id: number, name: string }>
store.userValue;  // Signal<string>
```

The behavior of `linkedSignal` and `resource` is preserved. Since the
SignalStore no longer creates the signals internally in these cases,
signals passed into `withState` can also be changed externally.

This is a foundational change to enable features like `withLinkedState`
and potential support for `withResource`.

The internal `STATE_SOURCE` is no longer represented as a single
`WritableSignal` holding the entire state object. Instead, each top-level
property becomes its own `WritableSignal`, or remains as-is if the user
already provides a `WritableSignal`.

## Motivation

- Internal creation of signals limited flexibility; users couldn’t bring
their own signals into the store
- Reusing existing signals enables future features like `withLinkedState`
or `withResource`.
- Splitting state into per-key signals improves the performance, because
the root is not the complete state anymore.

## Change to `STATE_SOURCE`

Given:

```ts
type User = {
  firstname: string;
  lastname: string;
};
```

### Before

```ts
STATE_SOURCE: WritableSignal<User>;
```

### Now

```ts
STATE_SOURCE: {
  firstname: WritableSignal<string>;
  lastname: WritableSignal<string>;
};
```

## Breaking Changes

### 1. Different object reference

The returned object from `signalState()` or `getState()` no longer keeps
the same object identity:

```ts
const obj = { ngrx: 'rocks' };
const state = signalState(obj);
```

**Before:**

```ts
state() === obj; // ✅ true
```

**Now:**

```ts
state() === obj; // ❌ false
```

---

### 2. No signal change on empty patch

Empty patches no longer emit updates, since no signal is mutated:

```ts
const state = signalState({ ngrx: 'rocks' });

let count = 0;
effect(() => count++);

TestBed.flushEffects();
expect(count).toBe(1);

patchState(state, {});
```

**Before:**

```ts
expect(count).toBe(2); // triggered
```

**Now:**

```ts
expect(count).toBe(1); // no update
```

---

### 3. No wrapping of top-level `WritableSignal`s

```ts
const Store = signalStore(
  withState({ foo: signal('bar') })
);
const store = new Store();
```

**Before:**

```ts
store.foo; // Signal<Signal<string>>
```

**Now:**

```ts
store.foo; // Signal<string>
```

---

### 4.: `patchState` no longer supports `Record` as root state

Using a `Record`as the root state is no longer supported by `patchState`.

**Before:**

```ts
const Store = signalStore(
  { providedIn: 'root' },
  withState<Record<number, number>>({}),
  withMethods((store) => ({
    addNumber(num: number): void {
      patchState(store, {
        [num]: num,
      });
    },
  }))
);

store.addNumber(1);
store.addNumber(2);

expect(getState(store)).toEqual({ 1: 1, 2: 2 });
```

**After:**

```ts
const Store = signalStore(
  { providedIn: 'root' },
  withState<Record<number, number>>({}),
  withMethods((store) => ({
    addNumber(num: number): void {
      patchState(store, {
        [num]: num,
      });
    },
  }))
);

store.addNumber(1);
store.addNumber(2);

expect(getState(store)).toEqual({}); // ❌ Nothing updated
```

If dynamic keys are needed, consider managing them inside a nested signal instead.

## Further Changes

- `signalStoreFeature` updated due to changes in `WritableStateSource`
- `patchState` now uses `NoInfer` on `updaters` to prevent incorrect type
  inference when chaining
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngrx/platform/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [] Documentation has been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Closes #

## What is the new behavior?

## Does this PR introduce a breaking change?

```
[x] Yes
[ ] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

markostanimirovic commented:
@rainerhahnekamp It's also necessary to write `BREAKING CHANGES: ...` in a plain text format before the `Other Information` section. It will be copied to the commit body on squash merge. See the example here: https://github.com/ngrx/platform/pull/4584

rainerhahnekamp commented:
> It's also necessary to write BREAKING CHANGES: ... 

Got it: Plural and then a line break.

It looks like my IDE messed up the subject a little bit. Will check that one as well.

rainerhahnekamp commented on modules/signals/src/state-source.ts:
```diff
@@ -30,40 +32,70 @@ export type StateWatcher<State extends object> = (
+  const stateKeys = Reflect.ownKeys(stateSource[STATE_SOURCE]);
+  for (const key of Reflect.ownKeys(newState)) {
+    if (!stateKeys.includes(key)) {
+      // TODO: Optional properties which don't exist in the initial state will not be added
```
We have here another breaking change. If we have a state with an optional root property and the initial state skips that one, `patchState` will always skip that one.

We could do a check for `undefined` in `withState` and/or come with an ESLint rule.

I've added some additional tests.

markostanimirovic replied:
Currently, it's also not 'allowed' to have optional state slices. If a state slice is not provided to the initial state, the state signal is never created.

rainerhahnekamp commented on modules/signals/src/with-state.ts:
```diff
@@ -35,21 +40,23 @@ export function withState<State extends object>(
+      stateSource[key] = isWritableSignal(signalValue)
+        ? signalValue
+        : signal(signalValue);
+      stateSignals[key] = toDeepSignal(stateSource[key]);
```
Marko, don't we have to assign them twice? Once for `STATE_SOURCE` and once for the state signals? That would be variables `stateSource` and `stateSignals`.

rainerhahnekamp commented on modules/signals/src/signal-store-feature.ts:
```diff
@@ -369,13 +369,13 @@ export function signalStoreFeature<
-    typeof featureOrInput === 'function'
-      ? [featureOrInput, ...restFeatures]
-      : restFeatures;
+  ...args:
+    | [Partial<SignalStoreFeatureResult>, ...SignalStoreFeature[]]
+    | SignalStoreFeature[]
+): SignalStoreFeature<EmptyFeatureResult, EmptyFeatureResult> {
+  const features = (
+    typeof args[0] === 'function' ? args : args.slice(1)
+  ) as SignalStoreFeature[];
```
With the change for the types, the typings stopped working for `signalStoreFeature` (not for `signalStore`). I had to apply this change to make it work again.

markostanimirovic replied:
I'll take a closer look

rainerhahnekamp commented on modules/signals/src/signal-state.ts:
```diff
@@ -1,18 +1,48 @@
+  const stateKeys = Reflect.ownKeys(initialState);
+  const stateAsRecord = initialState as Record<string | symbol, unknown>;
+
+  // define STATE_SOURCE property
```
Done

rainerhahnekamp commented on modules/signals/src/signal-state.ts:
```diff
@@ -1,18 +1,48 @@
-  const stateSource = signal(initialState as State);
-  const signalState = toDeepSignal(stateSource.asReadonly());
+  const stateKeys = Reflect.ownKeys(initialState);
+  const stateAsRecord = initialState as Record<string | symbol, unknown>;
```
Done for `stateAsRecord`. `stateKeys` is used three times.

rainerhahnekamp commented:
@markostanimirovic, I've updated the code or - where applicable - answered your comments. Please check, once you have time.

I've also fixed a bug in af974f9b.

I see commits from main have been merged. Can I rebase them instead or does GitHub do that automatically meanwhile?

markostanimirovic commented on modules/signals/src/state-source.ts:
```diff
@@ -30,40 +32,70 @@ export type StateWatcher<State extends object> = (
   stateSource: WritableStateSource<State>,
   ...updaters: Array<
-    Partial<Prettify<State>> | PartialStateUpdater<Prettify<State>>
+    | Partial<Prettify<NoInfer<State>>>
+    | PartialStateUpdater<Prettify<NoInfer<State>>>
```
Thanks for the explanation. I'll take a look

markostanimirovic commented:
> @markostanimirovic, I've updated the code or - where applicable - answered your comments. Please check, once you have time.
> 
> I've also fixed a bug in [af974f9](https://github.com/ngrx/platform/commit/af974f9b9f408192af2b467fa899d8888adce53d).
> 
> I see commits from main have been merged. Can I rebase them instead or does GitHub do that automatically meanwhile?

You can sync changes from `main` manually and push the changes, or use the "Update branch" button which is available on the PR page.

Btw, lint is also failing.

rainerhahnekamp commented on modules/signals/src/signal-state.ts:
```diff
@@ -1,18 +1,48 @@
 
 export function signalState<State extends object>(
   initialState: State
 ): SignalState<State> {
```
Done.

Implementation Detail: I moved the type `StateResult` into the `SignalState` instead of using it as return type of `signalState`. That has the advantage that the type `SignalState<State>` can be kept.

rainerhahnekamp commented:
@markostanimirovic, @timdeschryver 

- `signalState` now also support user-defined writable Signals
- I've updated the docs, by adding an example in `signalState` for both old and new website.

I will push the PR for `withLinkedState` later today.

timdeschryver commented on modules/signals/src/state-source.ts:
```diff
@@ -10,16 +10,26 @@ import {
+  [STATE_SOURCE]: { [Property in keyof State]: Signal<State[Property]> };
+};
+
+export type StateResult<StateInput extends object> = {
```
I wonder if this this should also be exported.

rainerhahnekamp replied:
I’m not sure. What criteria do you usually follow for exposing types?
If I don’t see an immediate need, I tend to hold off and wait for a concrete use case.

michael-small commented on projects/www/src/app/pages/guide/signals/signal-state.md:
````diff
@@ -54,6 +54,35 @@ console.log(firstName()); // logs: 'Eric'
+This allows an integration of external `WritableSignal`s — such as `linkedSignal` or `resource.value`.
+
+```ts
+import { linkedSignal } from '@angular/core';
+import { signalState } from '@ngrx/signals';
````
```suggestion
import { linkedSignal, signal } from '@angular/core';
import { patchState, signalState } from '@ngrx/signals';
```

timdeschryver commented on projects/www/src/app/pages/guide/signals/signal-state.md:
````diff
@@ -54,6 +54,35 @@ console.log(firstName()); // logs: 'Eric'
 console.log(lastName()); // logs: 'Clapton'
 ```
 
+If the root properties of a state are already of type `WritableSignal`, they will be reused, instead of creating new signals.
+This allows an integration of external `WritableSignal`s — such as `linkedSignal` or `resource.value`.
````
I think this one needs to be committed as well.

markostanimirovic commented on projects/ngrx.io/content/guide/signals/signal-state.md:
````diff
@@ -54,6 +54,20 @@ console.log(firstName()); // logs: 'Eric'
 console.log(lastName()); // logs: 'Clapton'
 ```
 
+When a state property holds an object as its value, the `signalState` function generates a `DeepSignal`.
+It can be used as a regular read-only signal, but it also contains signals for each property of the object it refers to.
+
+```ts
+const firstName = user.firstName; // type: Signal<string>
+const lastName = user.lastName; // type: Signal<string>
+
+console.log(firstName()); // logs: 'Eric'
+console.log(lastName()); // logs: 'Clapton'
+```
+
+If the root properties of a state are already a `WritableSignal`, then they are reused instead of creating new signals.
+This allows the integration of external `WritableSignal`s, such as `linkedSignal` or `resource.value`.
+
````
The previous paragraph and example are duplicated here accidentally. The example for the new `signalState` behavior is missing.

Let's leave 'Creating a SignalState' section as it is, and after 'Custom State Updaters' (before 'Usage') add a new section. Note: the new section should be added after 'Updating State' section because the example uses `patchState`. I also suggest not mentioning implementation details, like - "they are reused instead of creating new signals".

```md
## Using User-Defined Signals

The `signalState` function supports integrating user-defined `WritableSignal` instances as part of the state.

If a root state property is a `WritableSignal`, its value becomes an integral part of the state.
The SignalState instance and the original signal remain synchronized - updating one will immediately reflect in the other.

\```ts
const name = signal('Jimi Hendrix');
const userState = signalState({
  // 👇 Providing an external signal as part of the initial state.
  name,
  isAdmin: false,
});

console.log(userState.name()); // logs: Jimi Hendrix

// Updating the external signal
name.set('Brian May');
// reflects the value in the SignalState instance.
console.log(userState.name()); // logs: Brian May

// Updating the SignalState instance
patchState(userState, { name: 'Eric Clapton' });
// reflects the value in the external signal.
console.log(name()); // logs: Eric Clapton
\```
```

We should also add an alert/paragraph with similar content for `withState`.

rainerhahnekamp replied:
Done

rainerhahnekamp commented:
@markostanimirovic: I think we can start a new review round. There weren't actually that many changes:
- Applied your suggestions
- Added two tests for `withState`
- Removed the "equal check" in `patchState`
- Added an alert infobox about user-defined Signals for `withState`

timdeschryver commented on modules/signals/src/state-source.ts:
```diff
@@ -30,40 +40,83 @@ export type StateWatcher<State extends object> = (
+    const signalKey = key as keyof State;
+
+    if (currentState[signalKey] === newState[signalKey]) {
+      // continue;
```
I think this can be removed

markostanimirovic commented on projects/ngrx.io/content/guide/signals/signal-store/index.md:
```diff
@@ -49,6 +49,27 @@ The `BookSearchStore` instance will contain the following properties:
 
 <div class="alert is-helpful">
 
+Similar to `signalState`, `withState` accepts user-defined `WritableSignal` 
+instances as part of the state.
```
Let's add a bit more details here (assuming that reader is not familiar with `signalState`):

```suggestion
In addition to plain values, the `withState` feature also accepts user-defined `WritableSignal` instances as part of the state.
In this case, the SignalStore's state and the original signal remain synchronized - updating one will immediately reflect in the other.
```

markostanimirovic commented on projects/ngrx.io/content/guide/signals/signal-store/index.md:
````diff
@@ -49,6 +49,27 @@ The `BookSearchStore` instance will contain the following properties:
+instances as part of the state.
+
+```ts
+import { signalStore, withState } from '@ngrx/signals';
+import { Book } from './book';
+
+const booksSignal = signal<Book[]>([]);
+
+export const BookSearchStore = signalStore(
+  withState({
+    books: booksSignal,
+    isLoading: false
+  })
+);
````
and an example that shows how it works (note: signal suffix is removed from the variable name):

```suggestion
import { signalStore, withState } from '@ngrx/signals';
import { Book } from './book';

const books = signal<Book[]>([]);

const BookSearchStore = signalStore(
  withState({
    // 👇 Providing an external signal as part of the initial state.
    books,
    isLoading: false,
  })
);
```

markostanimirovic commented on modules/signals/spec/state-source.spec.ts:
```diff
@@ -145,6 +159,95 @@ describe('StateSource', () => {
+      });
+    });
+
+    it("sets all root proprerites and relies on the Signal's equals function", () => {
```
```suggestion
    it("sets all root properties and relies on the Signal's equals function", () => {
```
